c++/src/capnp/compat/json.c++

```diff
@@ -379,13 +379,119 @@ void JsonCodec::encodeField(StructSchema::Field field, DynamicValue::Reader inpu
   encode(input, field.getType(), output);
 }
 
+namespace {
+int64_t parseInt64(kj::StringPtr s)
+{
+  char *endPtr;
+  errno = 0;
+  int64_t value = std::strtoll(s.begin(), &endPtr, 10);
+  KJ_REQUIRE(endPtr == s.end() && errno != ERANGE, "String is not correct int64 number");
+  return value;
+}
+
+uint64_t parseUInt64(kj::StringPtr s)
+{
+  char *endPtr;
+  errno = 0;
+  uint64_t value = std::strtoull(s.begin(), &endPtr, 10);
+  KJ_ASSERT(endPtr == s.end() && errno != ERANGE, "String is not correct uint64 number");
+  return value;
+}
+
+template <typename Set, typename DecodeArray, typename DecodeObject>
+void decodeField(Type type, JsonValue::Reader value, Set set, DecodeArray decodeArray,
+    DecodeObject decodeObject) {
+  switch (value.which()) {
```

Comment:
I think the switch should be on the `type` rather than the `value` discriminant. This makes it easier to ensure that schema type is matching the JSON value type.